boards: arduino_nano_33_ble: add pyocd support (#51535)
zephyrproject-rtos/zephyr: boards/arm/arduino_nano_33_ble/doc/index.rst

You should see the the red LED blink.

Debugging
=========

You can debug an application on the board with a debug adapter that supports
CMSIS-DAP. This board has the SWD connector for debugging but exposes it as

> a test pad pattern (not a connector) on the back side of the PCB. So, It needs
> to be a bit difficult soldering. At least, the SWDIO and the SWCLK need
> to soldering. (As shown in the picture.)
> GND, 3.3V, and RESET are also available in the DIP connector. Therefore, it may be
> easier to connect these.

Reviewer: ```suggestion
bit of difficult soldering. At a minimum, SWDIO and SWCLK need soldering (As
shown in the picture). GND, 3.3V, and RESET are also available in the DIP
connector, therefore it may be easier to connect using the DIP connector
instead of soldering to them.
```

Author: Applied it.
Thank you for correcting.

.. image:: img/nano_33_ble_swd.jpg
     :align: center
     :alt: Nano 33 BLE SWD connecting

After connecting the debug adapter, you can debug it the usual way.
Type the following command will start debugging.

.. zephyr-app-commands::
   :zephyr-app: samples/basic/blinky
   :board: arduino_nano_33_ble
   :maybe-skip-config:

>    :goals: debug

Reviewer: Only if pyocd is the default debugger, otherwise it will need to be specified. The top runner in the board.cmake file is the default, which in this case would be bossac

Reviewer: This seems to be a bit of a confusing layout, e.g. the samba bootloader is specified in the dts, fine. But if you build an application with mcuboot and then flash it, bossac will fail (I would assume) since it would try to replace the bootloader using the bootloader, however if you did `west debug` it would replace the bootloader with mcuboot. So is the intention that the original bootloader would be used and the only bootloader, or would it want to be replaced with mcuboot, or would it want to support having the original bootloader load mcuboot which then loads the application?

Author: @nordicjm 

Thank you for commenting.

> This seems to be a bit of a confusing layout, e.g. the samba bootloader is specified in the dts, fine. But if you build an application with mcuboot and then flash it, bossac will fail (I would assume) since it would try to replace the bootloader using the bootloader, however if you did `west debug` it would replace the bootloader with mcuboot.

I tried to flash the bootloader with the bossac, and the command of bossac works, but it does not overwrite the bootloader. (It is a good design from the viewpoint of "fool-proof", I think.)

> So is the intention that the original bootloader would be used and the only bootloader, or would it want to be replaced with mcuboot, or would it want to support having the original bootloader load mcuboot which then loads the application?

In my opinion, the bootloader should use a user-specified one, and the system should not restrict it. And any bootloaders must be usable, but no need for any bootloader to work with all west commands.
In addition, it is better able to operate without specifying options in the state at the time of shipment.

From this point of view, the current behavior is preferable.
・Flashing ROM with the bossac command defaultly.
・It did not overwrite the bootloader by default, even flash using a debug adapter.
・Overwrite the boot loader only when flashing the boot loader is explicitly specified with the dts overlay file and runner option to the command line. (In this case, even if the device breaks, users can repair it themselves because the user has a debug adapter.)

To overwrite the bootloader, You need to change the `zephyr,code-partition` settings in dts. `west debug` or `west flash`  doesn't overwrite the bootloader unless you overwrite the setting.
(The boot/zephyr/ sample in the mcuboot repository is an example of rewriting the bootloader. Therefore, specifying boot-partition to the `zephyr,code-partition` as in the dts.overlay below is necessary.
And the bootloader needs to flash separately from the application.
https://github.com/zephyrproject-rtos/mcuboot/blob/main/boot/zephyr/dts.overlay)

Also, as mentioned above, the bossac cannot overwrite the bootloader, so connecting a separate debug adapter is necessary. It means that the user must explicitly specify bootloader rewrites. (In other words, it won't break unless you don't do anything extra.)

Please mention if my understanding is wrong.
